### PR TITLE
`File upload exercises`: Allow upload of multiple files

### PR DESCRIPTION
**Pull request by theblobinthesky**
> [!WARNING]
> ### Database changes not yet approved for deployment on test servers.

#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
During an exam at the University of Passau some students had issues uploading their work in the file upload exercise. Since it only accepts one image or pdf at a time and their work was distributed on multiple pages time issues arose.
To prevent this problem in the future this feature handles uploading multiple files rather than one file per file upload exercise.

### Description
Rather than uploading and submitting a single file the client can upload a list of files, delete files from the list again and submit them as usual. This works in exam mode and viewing the multiple files works in feedback mode as well.

### Steps for Testing
Steps to test uploading multiple files in a regular file upload exercise:
Prerequisites:
- 1 Instructor
- 1 Student
- 1 File Upload Exercise

1. Log in to Artemis as the student.
2. Navigate to Course Overview, start and upload file in file upload exercise.
3. Click on the file selector button and select a file with a supported extension.
4. Acknowledge the file was added to the "Files to be submitted once you're ready" stage list, the entry opens up the file when clicked and the submit button is not disabled.
5. Delete the file and acknowledge it is gone from the list and the submit button is disabled.
6. Select two more files, acknowledge both entries open up the right file when clicked and submit them.
7. Acknowledge the stage list is gone and the submissions are listed under "Submitted files as solution for the exercise {time}" in the right order and that the entries download the right files when clicked.
8. Log in to Artemis as the instructor.
9. Navigate to Course Management and open the submissions assessment window.
10. Make sure clicking the file entries downloads the right files.
11. Submit an assessment.
12. Log in to Artemis as the student.
13. Review the exercise and check the file entries download the right files.

Steps to test exports of multiple files:
Prerequisites:
- 1 Instructor
- 1 Student
- 1 File Upload Exercise

1. Log in to Artemis as the student and submit multiple files in the file upload exercise as described in the paragraph above.
2. Log in to Artemis as the instructor and open the file upload exercises scores panel.
3. Click "Export", export the submissions of the student and save the archive.
4. Acknowledge the downloaded archive contains an archive named "{exercise name}-{student name}-{submission}.zip". This archive should then contain the submission files.

#### Exam Mode Testing
Steps to test uploading multiple files in exam mode:
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a File Upload Exercise

1. Log in to Artemis as the student.
5. Participate in the exam as a student and open the file upload exercise.
6. Select two files, acknowledge both entries open up the right file when clicked and click the upload button.
7. Acknowledge the stage list is gone and the submissions are listed under "Submitted files as solution for the exercise {time}" in the right order and that the entries download the right files when clicked.
8. Log in to Artemis as the instructor. 
9. Assess the exam and check the entries download the right files as before.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Screenshots
Showcasing uploading multiple files in a regular file upload exercise:

[Showcasing uploading multiple files in a regular file upload exercise](https://github.com/ls1intum/Artemis/assets/53352472/2305e3ee-4eb3-4060-9980-275b082680ef)

Showcasing uploading multiple files in exam mode:

[Showcasing uploading multiple files in exam mode](https://github.com/ls1intum/Artemis/assets/53352472/62ebb75f-1c13-4ca5-b844-e607aba25fbb)

Showcasing exports of multiple files:

[Showcasing exports of multiple files](https://github.com/ls1intum/Artemis/assets/53352472/9251496d-5300-4d10-a1ae-96585b98415a)

### CODE REVIEW

**Comment by b-fein** [comment]
Thanks for contributing this feature!

---

> The latest develop commit that has been merged into this branch in the last commit ([Merged develop], https://github.com/ls1intum/Artemis/commit/d5f0f1ebc47e887868ac9cb8bc462b2b38cd9f0a) does not work on my machine since the spring application fails to startup on the latest develop branch commit on my PC.

Two questions to make debugging easier:
- How do you start Artemis (IntelliJ run configuration, Docker, …)?
- What error message/stack trace do you get?

**Comment by theblobinthesky** [comment]
Thanks for responding- for debugging I've been running an instance of docker/mysql.yml in docker-compose and the "Artemis (Server & Client)" run configuration in Intellij. I don't run jira currently because i thought you needed a key for that. Since i'm not using programming exercises Bitbucket and Bamboo should be optional for me right?

**Comment by b-fein** [comment]
> Since i'm not using programming exercises Bitbucket and Bamboo should be optional for me right?

Yes. You might need to remove the `bamboo,bitbucket,jira,` profiles from the ‘Active Profiles’ string in the run configurations.

I think what happened before is that those were effectively ignored as long as you did not create a programming exercise in the system. However, in 6.5.0 a migration was introduced that needs a connection to those external systems *if* one of those profiles is active. So, by removing the profiles the (in your case unnecessary) migration will be skipped and Artemis should start again.
If that does not work, please post the error message and stack trace Artemis crashes with.

**Comment by theblobinthesky** [comment]
Thanks for the suggestion, that solved my problem. The newest commit seems to work as expected.

**Comment by b-fein** [comment]
Perfect!

- Looks like the client tests are failing currently due to type errors (you can scroll all the way down on https://github.com/ls1intum/Artemis/pull/7268/files).
- Some server tests seem to be failing, too. (`ExamIntegrationTest` now has a ZIP for the files instead of a single image, `CourseGitlabJenkinsIntegrationTest::testDownloadCourseArchiveAsInstructor`). GitHub actions sadly does not report them directly here when running on a forked project.
- For `if`/`for`/... the code style usually requires braces (`FileUploadSubmission.java`), not sure why the autoformatter did not fix those in your case.

After you’ve had a look at the failing tests, make sure to click the ‘Ready for review’ button below the comments in the pull request when you think it’s ready for code review and manual testing.

File: .idea/runConfigurations/Artemis__Server_.xml [hunk 1]
```diff
@@ -1,13 +1,12 @@
 <component name="ProjectRunConfigurationManager">
```
Comment (Strohgelaender) [inline .idea/runConfigurations/Artemis__Server_.xml:1]: Can you please revert the changes in this file (and keep them only local) so that other developers still have the bamboo / bitbucket / jira profiles here.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 2]
```diff
@@ -290,7 +291,14 @@ public ResponseEntity<byte[]> getFileUploadSubmission(@PathVariable Long exercis
             throw new AccessForbiddenException();
         }
 
-        return buildFileResponse(filePathService.actualPathForPublicPath(URI.create(submission.getFilePath())), false);
+        String filePath = submission.getFilePaths().stream().filter(path -> path.endsWith(fileName)).findFirst().orElse(null);
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:294]: Have you tested what would happen if a student uploads two different files with the same name? Are they correctly saved and retrieved?

File: src/main/resources/config/liquibase/changelog/20230906184000_changelog.xml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
+    <changeSet author="erik" id="20230812135300">
+        <comment>
+            Create the table to store multiple paths with a file submission
+            instead of only one and copy the old paths into the new table.
+        </comment>
+        <createTable tableName="file_submission_paths">
+            <column name="submission_id" type="bigint">
+                <constraints nullable="false" />
```
Comment (b-fein) [inline src/main/resources/config/liquibase/changelog/20230906184000_changelog.xml:12]: Should this also have a foreign key constraint pointing to the `id` column of the `submission` table?

File: src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java [hunk 4]
```diff
@@ -25,21 +26,44 @@ public String getSubmissionExerciseType() {
     @Transient
     private final transient FileService fileService = new FileService();
 
-    @Column(name = "file_path")
-    private String filePath;
+    @ElementCollection(fetch = FetchType.EAGER)
+    @CollectionTable(name = "file_submission_paths", joinColumns = @JoinColumn(name = "submission_id"))
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java:30]: Does this already imply proper cascading to the file paths table? I.e., are the rows connected to a submission deleted when a submission is deleted?

File: src/main/webapp/app/exercises/file-upload/stage/file-upload-stage.component.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
+import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
+import { MAX_SUBMISSION_FILE_SIZE } from 'app/shared/constants/input.constants';
+import { FileDetails } from 'app/entities/file-details.model';
+import { FileUploadSubmission } from 'app/entities/file-upload-submission.model';
+import { AlertService } from 'app/core/util/alert.service';
+import { FileService } from 'app/shared/http/file.service';
+
+@Component({
+    selector: 'jhi-stage',
```
Comment (b-fein) [inline src/main/webapp/app/exercises/file-upload/stage/file-upload-stage.component.ts:9]:
```suggestion
    selector: 'jhi-file-upload-stage',
```
A more specific name would be useful to understand what this component is in places where it is used without having to look at further details.

File: src/main/webapp/app/entities/file-details.model.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+export class FileDetails {
+    constructor(
+        public name: string = '',
+        public extension: string = '',
+    ) {}
```
Comment (b-fein) [inline src/main/webapp/app/entities/file-details.model.ts:5]:
```suggestion
    ) {}

```
Nice refactoring of functionality that is used in multiple places!

File: src/main/webapp/app/entities/file-details.model.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    static getFileDetailsFromPath(filePath: string): FileDetails {
```
Comment (b-fein) [inline src/main/webapp/app/entities/file-details.model.ts:6]:
```suggestion
    static getFileDetailsFromPath(filePath: string | undefined): FileDetails {
```
You are checking in the next line if `filePath` is not there. The typing should reflect that this can be possible.

File: src/main/webapp/i18n/de/fileUploadSubmission.json [hunk 8]
```diff
@@ -24,8 +24,10 @@
             "fileExtensionError": "Falsche Dateierweiterung. Bitte lade eine Datei mit einer der unten aufgeführten zulässigen Dateierweiterungen hoch.",
             "fileUploadError": "Deine Einreichung der Datei {{ fileName }} ist fehlgeschlagen!",
             "submitted": "Abgegeben",
-            "submittedFile": "Die Datei {{ filename }} wurde als Lösung für die Übung eingereicht",
-            "download": "Datei herunterladen",
+            "stagedFiles": "Die Dateien die abgegeben werden sobeit du bereit bist",
```
Comment (b-fein) [inline src/main/webapp/i18n/de/fileUploadSubmission.json:27]:
```suggestion
            "stagedFiles": "Die Dateien die abgegeben werden sobald du bereit bist",
```

**Review by b-fein — COMMENTED** [review]
Overall, code looks good. I’ve added a few suggestions for smaller improvements and some questions.

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java [hunk 9]
```diff
@@ -815,7 +815,7 @@ private boolean hasNonEmptySubmission(Set<Submission> submissions, Exercise exer
         }
         else if (exercise instanceof FileUploadExercise) {
             FileUploadSubmission textSubmission = (FileUploadSubmission) submissions.iterator().next();
-            return textSubmission.getFilePath() != null && !textSubmission.getFilePath().isEmpty();
+            return !textSubmission.getFilePaths().isEmpty() && textSubmission.getFilePaths().stream().noneMatch(String::isEmpty);
```
[the remark this answers is not in the document]
Comment (theblobinthesky) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java:818]: I've checked this out and the fileUploadSubmission.getFilePaths() URI's come from FilePathService::publicPathForActualPath. I don't think this method has a path that returns an empty string but if the user client submits a file that starts with an unknown directory it will return null. Still I think it's a good idea to apply your suggestion just in case there might be empty strings for some reason. I would also filter out all null entries right where publicPathForActualPath is called in FileUploadSubmissionService::storeFiles since (especially) stream operations like map on null entries might cause an exception unexpectedly somewhere downstream when called on nulls.

File: src/main/java/de/tum/in/www1/artemis/service/export/FileUploadSubmissionExportService.java [hunk 10]
```diff
@@ -21,48 +21,43 @@
...
+        List<Path> filePaths = new ArrayList<>();
 
-        if (!Files.exists(filePath)) { // throw if submission file does not exist
-            throw new IOException("Cannot export submission " + submission.getId() + " because the uploaded file " + filePath + " doesn't exist.");
-        }
+        for (String urlFilePath : urlFilePaths) {
+            // we need to get the 'real' file path here, the submission only has the api url path
+            Path filePath = FileUploadSubmission.buildFilePath(exercise.getId(), submission.getId());
+            String[] apiFilePathParts = urlFilePath.replace("/", File.separator).split(Pattern.quote(File.separator));
+            String fileName = apiFilePathParts[apiFilePathParts.length - 1];
```
[the remark this answers is not in the document]
Comment (theblobinthesky) [reply src/main/java/de/tum/in/www1/artemis/service/export/FileUploadSubmissionExportService.java]: Tested and works.

File: src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java (hunk 4 above)
Comment (theblobinthesky) [reply src/main/java/de/tum/in/www1/artemis/domain/FileUploadSubmission.java:30]: Deleting a participation and a submission results in the submission to be removed from the submission table and all it's paths to be removed from the file_submission_paths table.

**Comment by theblobinthesky** [comment]
I've noticed on the latest develop commit the old file in the file upload submission is not deleted anymore. I think this bug was introduced here: https://github.com/ls1intum/Artemis/commit/aca8f44609161249ada52ede6028091c2096e2c0#diff-afbf7c4cfa2c701ea17f5d3fc9cae4f94e88f182921364471f35e9f3355e0302R37. 
The filePath in FileUploadSubmission and now the filePaths in FileUploadSubmission are public paths not the actual path on the filesystem. fileService.schedulePathForDeletion takes actual paths though. That means old submission files are never deleted and just accumulate in the submission directory. The fix is quite simple but the question is if i should include it here or if that should be seperate

**Comment by b-fein** [comment]
> The fix is quite simple but the question is if i should include it here or if that should be seperate

I think opening an issue for that so that a separate small bugfix-PR can be created (either by you or someone else) and merged quickly would be good.

**Comment by theblobinthesky** [comment]
I've created an issue https://github.com/ls1intum/Artemis/issues/7275 and submitted a pull request https://github.com/ls1intum/Artemis/pull/7276.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java (hunk 2 above)
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:294]: I suppose it’s quite the edge-case. If it is hard to implement, I guess forbidding multiple files having identical names would be okay, too.

Comment (theblobinthesky) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:294]: The most obvious solution which I've already implemented is to compute a hash of the file contents when they are staged in the stage component.
To continue displaying the correct filename without the hash prefix, you just modify the file details function to cut off the hash before displaying it.
That would not require any more testing on the server and client side since the change is isolated to two files.

Only problem is the database migration can't compute the hash.

Even more promising would be to prefix the file with its index in the list instead of a hash but that would require a small change to the server.
I'm looking into that now especially because you might be able to notice the delay caused by the hash function on some slow clients.

Long term you'd probably want to save the name in a seperate column and just refer to files by /api/path/{hash}.{extension} instead of /api/path/{name}.{extension} because that would leave the spaces, special chars etc in the names intact. But that would require much more significant change and should probably be a seperate pr at some point.

Comment (theblobinthesky) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:294]: Nevermind the index version should work. I got rid of the hash

**Comment by theblobinthesky** [comment]
This should be everything.

**Comment by krusche** [comment]
Does this PR also allow tutors to review multiple files so they can assess them and provide feedback?

Are there other places, e.g. for instructors, that need to be adapted when we allow multiple files?

Would it make sense for an instructor to specify how many files can be submitted? The default value could be 1 to preserve the current behavior...

**Comment by theblobinthesky** [comment]
**Does this PR also allow tutors to review multiple files so they can assess them and provide feedback?:**
It allows tutors to open all files in the submission as it is shown in the video at the bottom of the pr but there is no option to give feedback on a specific file.

**Are there other places, e.g. for instructors, that need to be adapted when we allow multiple files?:**
As of now I've replaced the option to download the submission file with the option to download multiple files (so also as an instructor giving feedback) and replaced the option to upload a file with the option to upload multiple files in all the places i could find.

**Would it make sense for an instructor to specify how many files can be submitted?:**
The original goal of the pr was to avoid situations where students are forced to scramble mid-exam. In this case some students wrote one part of the question on the front side and the other on the back side of their paper. They were then forced to scramble and convert their pictures into a pdf mid-exam. Limiting to one file per-default would probably cause this problem to persist. A default limit >1 and the option to limit the amount of files could be useful for instructors though.

**Comment by theblobinthesky** [comment]
A comment to keep the pr alive
